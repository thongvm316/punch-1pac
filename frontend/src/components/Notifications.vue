<template>
  <div>
    <div
      ref="dropdownMenu"
      class="dropdown mr-5"
      :class="{ active: isDropdownActive }"
      @click="toggleDropdown"
    >
      <span
        class="notification"
        :class="{ badge: unreadNotificationsCount }"
        :data-badge="displayNotificationsCount"
      >
        <p-ico-bell />
      </span>
      <div class="box notifications notification-dropdown triangle-top">
        <div class="notification-header">
          <h4>{{ $t('header.notifications') }}</h4>
        </div>
        <ul
          v-if="headerNotifications.length"
          ref="notiList"
        >
          <li
            v-for="headerNotification in headerNotifications"
            :key="headerNotification.id"
            :class="{ 'notification-pending': headerNotification.activitable.status === 'pending' }"
            @click="openRequestModal(headerNotification)"
          >
            <div class="tile tile-centered tile-activity">
              <div class="tile-icon">
                <img
                  :src="headerNotification.user.avatar_url"
                  class="avatar avatar-md"
                  :alt="headerNotification.user.name"
                >
              </div>
              <div class="tile-content">
                <p
                  class="tile-title"
                  v-html="$t(`notification.${headerNotification.activitable_type.toLowerCase()}.${headerNotification.kind}`, { name: headerNotification.user.name })"
                />
                <p class="tile-subtitle">
                  {{ headerNotification.created_at | moment_activity }}
                </p>
              </div>
              <div
                v-if="headerNotification.activitable.status === 'pending'"
                class="tile-action"
              >
                <span class="label label-warning">{{ $t('meta.request_statuses.pending') }}</span>
              </div>
            </div>
          </li>
        </ul>
        <p
          v-else
          class="no-notification-msg"
        >
          {{ $t('header.noNotificationMsg') }}
        </p>
      </div>
    </div>
    <modal
      v-if="isEditable(notification)"
      ref="requestModal"
      :title="isRequestDayOff ? $t('modal.annualLeave.title') : $t('modal.attendance.editTitle') "
      :modal-open.sync="isAddModalOpen"
    >
      <div class="form-group">
        <label class="form-label">{{ $t('label.date') }}</label>
        <flat-pickr
          :config="{ mode: 'single', enable: [notification.activitable.attendance_day], locale: flatpickrLocaleMapper[pickrLocale] }"
          class="form-input daterange-picker"
          :value="notification.activitable.attendance_day"
          disabled
        />
      </div>
      <div
        v-if="!isRequestDayOff"
        class="form-group"
      >
        <label class="form-label">{{ $t('label.attendedAt') }}</label>
        <input
          type="time"
          class="form-input time-picker"
          :value="notification.activitable.attended_at"
          disabled
        >
      </div>
      <div
        v-if="!isRequestDayOff"
        class="form-group"
      >
        <label class="form-label">{{ $t('label.leftAt') }}</label>
        <input
          type="time"
          class="form-input time-picker"
          :value="notification.activitable.left_at"
          disabled
        >
      </div>
      <div class="form-group">
        <label class="form-label">{{ $t('label.reason') }}</label>
        <textarea
          class="form-input"
          :value="notification.activitable.reason"
          disabled
        />
      </div>
      <div class="form-group">
        <label class="form-label">{{ $t('label.rejectReason') }}</label>
        <textarea
          v-model="rejectReason"
          class="form-input"
          :placeholder="$t('label.rejectReason')"
        />
      </div>
      <div class="modal-footer">
        <button
          type="button"
          class="btn btn-success"
          @click="submitAddModal(notification.activitable_id, approveNotificationRequest, $t('messages.request.approvedSuccess'))"
        >
          {{ $t('button.common.approve') }}
        </button>
        <button
          type="button"
          class="btn btn-error"
          @click="submitAddModal({ id: notification.activitable_id, admin_reason: rejectReason }, rejectNotificationRequest, $t('messages.request.rejectedSuccess'))"
        >
          {{ $t('button.common.reject') }}
        </button>
      </div>
    </modal>
  </div>
</template>

<script>
import { mapState, mapGetters, mapActions } from 'vuex'
import flatpickrLocale from '@/mixins/flatpickr-locale'
import dropdown from '@/mixins/dropdown'
import modal from '@/mixins/modal'
const PIcoBell = () => import('@/punch-ui/punch-icons/PIcoBell')
const flatPickr = () => import('vue-flatpickr-component')

export default {
  name: 'Notifications',

  components: {
    flatPickr,
    PIcoBell
  },

  mixins: [dropdown, modal, flatpickrLocale],

  data() {
    return {
      notification: {},
      rejectReason: ''
    }
  },

  computed: {
    ...mapState('notifications', ['unreadNotificationsCount', 'headerNotifications', 'pager']),

    ...mapGetters('notifications', ['displayNotificationsCount']),

    isRequestDayOff() {
      return this.notification.activitable.kind === 'annual_leave'
    }
  },

  created() {
    this.getHeaderNotifications()
  },

  methods: {
    loadMoreOnScroll() {
      let el = this.$refs.notiList

      if (el.scrollTop + el.offsetHeight === el.scrollHeight && this.pager.next_page) {
        this.getMoreHeaderNotifications({ page: this.pager.next_page })
      }
    },

    toggleDropdown() {
      let el = this.$refs.notiList

      this.isDropdownActive = !this.isDropdownActive
      if (this.isDropdownActive && this.headerNotifications[0]) this.readNotifications(this.headerNotifications[0].id)
      if (this.isDropdownActive && el) {
        el.addEventListener('scroll', this.loadMoreOnScroll)
      } else if (!this.isDropdownActive && el) {
        el.removeEventListener('scroll', this.loadMoreOnScroll)
      }
    },

    ...mapActions('notifications', [
      'readNotifications',
      'getHeaderNotifications',
      'getMoreHeaderNotifications',
      'approveNotificationRequest',
      'rejectNotificationRequest'
    ]),

    openRequestModal(notification) {
      this.isAddModalOpen = !this.isAddModalOpen
      this.notification = notification
    },

    isEditable(notification) {
      return ['create', 'update'].includes(notification.kind) && notification.activitable_type === 'Request' && !['approved', 'rejected'].includes(notification.activitable.status)
    }
  }
}
</script>
