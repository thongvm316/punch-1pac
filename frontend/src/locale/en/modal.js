export default {
  request: {
    addTitle: 'Add request',
    editTitle: 'Edit request',
    title: 'What kind of request do you want to send ?',
    kind: {
      attendance: 'Request for editing attendance',
      annual_leave: 'Request a day off'
    }
  },
  annualLeave: {
    title: 'Request a leave'
  },
  group: {
    addTitle: 'Add group',
    editTitle: 'Edit Group'
  },
  user: {
    editTitle: 'Edit User'
  },
  businessDays: {
    addTitle: 'Add business day',
    editTitle: 'Edit business day'
  },
  allowedIPs: {
    addTitle: 'Add IP address',
    editTitle: 'Edit IP address'
  },
  holiday: {
    addTitle: 'Add holiday',
    editTitle: 'Edit holiday'
  }
}
