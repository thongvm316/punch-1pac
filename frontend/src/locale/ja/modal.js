export default {
  request: {
    addTitle: '修正を依頼する',
    editTitle: '申請の編集',
    title: '何を申請しますか？',
    kind: {
      attendance: '勤怠の修正',
      annual_leave: '休暇・欠勤'
    }
  },
  annualLeave: {
    title: '休暇申請'
  },
  group: {
    addTitle: 'グループの追加',
    editTitle: 'グループ情報の編集'
  },
  user: {
    editTitle: 'ユーザー情報の編集'
  },
  businessDays: {
    addTitle: '営業日の追加',
    editTitle: '営業日の編集'
  },
  allowedIPs: {
    addTitle: 'IPアドレスの追加',
    editTitle: 'IPアドレスの編集'
  },
  holiday: {
    addTitle: '休日の追加',
    editTitle: '休日の編集'
  }
}
