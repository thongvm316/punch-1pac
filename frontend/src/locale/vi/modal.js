export default {
  request: {
    addTitle: 'Add request',
    editTitle: 'Chỉnh sửa yêu cầu',
    title: 'Loại yêu cầu bạn muốn gởi ?',
    kind: {
      attendance: 'Tạo yêu cầu hiện diện',
      annual_leave: 'Tạo yêu cầu xin nghỉ'
    }
  },
  annualLeave: {
    title: 'Xin nghỉ'
  },
  group: {
    addTitle: 'Tạo nhóm',
    editTitle: 'Sửa thông tin nhóm'
  },
  user: {
    editTitle: 'Sửa thông tin thành viên'
  },
  businessDays: {
    addTitle: 'Tạo ngày đi làm',
    editTitle: 'Chỉnh sửa ngày đi làm'
  },
  allowedIPs: {
    addTitle: 'Tạo địa chỉ IP',
    editTitle: 'Chỉnh sửa địa chỉ IP'
  },
  holiday: {
    addTitle: 'Tạo ngày lễ',
    editTitle: 'Chỉnh sửa ngày lễ'
  }
}
