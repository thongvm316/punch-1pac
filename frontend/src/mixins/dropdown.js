export default {
  data() {
    return {
      isDropdownActive: false,
      isLangSelectActive: false
    }
  },

  methods: {
    closeDropdown(e) {
      const el = this.$refs.dropdownMenu
      const target = e.target
      if (el !== target && !el.contains(target)) {
        this.isDropdownActive = false
        this.isLangSelectActive = false
      }
    }
  },

  created() {
    document.addEventListener('click', this.closeDropdown)
  },

  destroyed() {
    document.removeEventListener('click', this.closeDropdown)
  }
}
