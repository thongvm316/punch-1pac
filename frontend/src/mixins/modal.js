import Modal from '../components/Modal.vue'
import { mapActions } from 'vuex'

export default {
  data() {
    return {
      isAddModalOpen: false,
      isEditModalOpen: false
    }
  },

  components: {
    Modal
  },

  methods: {
    ...mapActions('flash', ['setFlashMsg']),

    toggleAddModal() {
      this.isAddModalOpen = !this.isAddModalOpen
    },

    toggleEditModal() {
      this.isEditModalOpen = !this.isEditModalOpen
    },

    submitAddModal(params, callback, msg = '') {
      callback(params).then(response => {
        this.isAddModalOpen = false
        if (msg) this.setFlashMsg({ message: msg })
      })
    },

    saveEditModal(params, callback, msg = '') {
      callback(params).then(response => {
        this.isEditModalOpen = false
        if (msg) this.setFlashMsg({ message: msg })
      })
    }
  }
}
