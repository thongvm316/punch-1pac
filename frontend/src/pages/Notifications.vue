<template>
  <main-layout :title="$t('notifications.title')">
    <div class="box notifications">
      <ul>
        <li v-for="notification in notifications">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon">
              <img :src="notification.user.avatar_url" class="avatar avatar-md" :alt="notification.user.name">
            </div>
            <div class="tile-content">
              <p class="tile-title" v-html="$t(`activity.${notification.activitable_type.toLowerCase()}.${notification.kind}`, { name: notification.user.name })"></p>
              <p class="tile-subtitle">{{ notification.created_at | moment_activity }}</p>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <pagination action="getNotifications" namespace="notifications" v-if="pager.total_pages > 1"/>
  </main-layout>
</template>

<script>
import MainLayout from '../layouts/Main.vue'
import Pagination from '../components/Pagination.vue'
import { mapState, mapActions } from 'vuex'

export default {
  components: {
    MainLayout,
    Pagination
  },

  computed: {
    ...mapState('notifications', [
      'pager',
      'notifications'
    ])
  },

  methods: {
    ...mapActions('notifications', [
      'getNotifications'
    ])
  },

  created () {
    this.getNotifications()
  }
}
</script>
