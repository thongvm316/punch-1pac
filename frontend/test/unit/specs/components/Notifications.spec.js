import { shallowMount } from '@vue/test-utils'

import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'

import flatpickrLocale from '@/mixins/flatpickr-locale'
import dropdown from '@/mixins/dropdown'
import modal from '@/mixins/modal'
import Notifications from '@/components/Notifications'

const headerNotifications = [
  {
    id: 0,
    activitable: {
      status: 'pending',
      type: 'Request'
    },
    user: {},
    activitable_type: 'Request',
    kind: 'update',
    create_at: '2019-12-02'
  },
  {
    id: 1,
    activitable: {
      status: 'approved'
    },
    user: {},
    kind: 'update',
    activitable_type: 'Request',
    create_at: '2019-12-04'
  }
]
const getHeaderNotifications = jest.fn()
const toggleDropdown = jest.spyOn(Notifications.methods, 'toggleDropdown')
const isEditable = jest.spyOn(Notifications.methods, 'isEditable')
const loadMoreOnScroll = jest.fn()
const readNotifications = jest.fn()

Object.assign(wrapperOps, {
  methods: {
    getHeaderNotifications,
    toggleDropdown,
    loadMoreOnScroll,
    isEditable,
    readNotifications
  },
  mixins: [dropdown, modal, flatpickrLocale]
})

describe('Notifications.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Notifications, wrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when Notifications was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
    })
  })

  describe('when click dropdownMenu', () => {
    it('should call toggleDropdown method', async () => {
      const dropdownMenu = wrapper.find({ ref: 'dropdownMenu' })
      dropdownMenu.trigger('click')
      await wrapper.vm.$nextTick()

      expect(toggleDropdown).toHaveBeenCalled()
      expect(wrapper.vm.isDropdownActive).toBe(true)
      expect(dropdownMenu.isVisible()).toBeTruthy()
    })
  })

  describe('when dont have headerNotifications', () => {
    it('should render no header notifications', async () => {
      setComputed(wrapper, { headerNotifications: [] })
      await wrapper.vm.$nextTick()

      expect(wrapper.vm.headerNotifications).toHaveLength(0)
      expect(wrapper.find({ ref: 'notiList' }).exists()).toBeFalsy()
    })
  })

  describe('when have headerNotifications', () => {
    let editableNoti, uneditableNoti

    beforeEach(async () => {
      setComputed(wrapper, { headerNotifications })
      await wrapper.vm.$nextTick()

      editableNoti = wrapper.find({ ref: 'notiList' }).find('li:first-child')
      uneditableNoti = wrapper.find({ ref: 'notiList' }).find('li:last-child')
    })

    it('should render header notifications', () => {
      expect(wrapper.vm.headerNotifications).toHaveLength(2)
      expect(wrapper.find({ ref: 'notiList' }).exists()).toBeTruthy()
    })

    it('editable noti should show pending label', () => {
      expect(editableNoti.find('.label-warning').exists()).toBeTruthy()
    })

    it('should show modal when click on editable request notification', async () => {
      editableNoti.trigger('click')
      await wrapper.vm.$nextTick()

      expect(wrapper.find({ ref: 'requestModal' }).exists()).toBeTruthy()
      expect(wrapper.find({ ref: 'requestModal' }).isVisible()).toBeTruthy()
    })

    it('uneditable noti should show pending label', () => {
      expect(uneditableNoti.find('.label-warning').exists()).toBeFalsy()
    })

    it('modal should not exist when click on uneditable request notification', async () => {
      uneditableNoti.trigger('click')
      await wrapper.vm.$nextTick()

      expect(wrapper.find({ ref: 'requestModal' }).exists()).toBeFalsy()
    })

    it('should call loadMoreOnScroll when user open noti list and trigger scroll', async () => {
      wrapper.find({ ref: 'dropdownMenu' }).trigger('click')
      wrapper.find({ ref: 'notiList' }).trigger('scroll')
      await wrapper.vm.$nextTick()

      expect(loadMoreOnScroll).toHaveBeenCalled()
    })
  })
})
