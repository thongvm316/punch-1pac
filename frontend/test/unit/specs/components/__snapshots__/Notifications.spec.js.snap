// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications.vue when Notifications was mounted should render correctly 1`] = `
<div>
  <div class="dropdown mr-5 active"><span data-badge="0" class="notification"><picobell-stub></picobell-stub></span>
    <div class="box notifications notification-dropdown triangle-top">
      <div class="notification-header">
        <h4>Notifications</h4>
      </div>
      <p class="no-notification-msg">
        You have no new notifications
      </p>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Notifications.vue when have headerNotifications should not exist modal when click on uneditable request notification 1`] = `
<div>
  <div class="dropdown mr-5"><span data-badge="0" class="notification"><picobell-stub></picobell-stub></span>
    <div class="box notifications notification-dropdown triangle-top">
      <div class="notification-header">
        <h4>Notifications</h4>
      </div>
      <ul>
        <li class="notification-pending">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon"><img class="avatar avatar-md"></div>
            <div class="tile-content">
              <p class="tile-title"><span class="text-bold-600"></span> updated a request</p>
              <p class="tile-subtitle">

              </p>
            </div>
            <div class="tile-action"><span class="label label-warning">Pending</span></div>
          </div>
        </li>
        <li class="">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon"><img class="avatar avatar-md"></div>
            <div class="tile-content">
              <p class="tile-title"><span class="text-bold-600"></span> updated a request</p>
              <p class="tile-subtitle">

              </p>
            </div>
            <!---->
          </div>
        </li>
      </ul>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Notifications.vue when have headerNotifications should render header notifications 1`] = `
<div>
  <div class="dropdown mr-5 active"><span data-badge="0" class="notification"><picobell-stub></picobell-stub></span>
    <div class="box notifications notification-dropdown triangle-top">
      <div class="notification-header">
        <h4>Notifications</h4>
      </div>
      <ul>
        <li class="notification-pending">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon"><img class="avatar avatar-md"></div>
            <div class="tile-content">
              <p class="tile-title"><span class="text-bold-600"></span> updated a request</p>
              <p class="tile-subtitle">

              </p>
            </div>
            <div class="tile-action"><span class="label label-warning">Pending</span></div>
          </div>
        </li>
        <li class="">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon"><img class="avatar avatar-md"></div>
            <div class="tile-content">
              <p class="tile-title"><span class="text-bold-600"></span> updated a request</p>
              <p class="tile-subtitle">

              </p>
            </div>
            <!---->
          </div>
        </li>
      </ul>
    </div>
  </div>
  <!---->
</div>
`;

exports[`Notifications.vue when have headerNotifications should show modal when click on editable request notification 1`] = `
<div>
  <div class="dropdown mr-5"><span data-badge="0" class="notification"><picobell-stub></picobell-stub></span>
    <div class="box notifications notification-dropdown triangle-top">
      <div class="notification-header">
        <h4>Notifications</h4>
      </div>
      <ul>
        <li class="notification-pending">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon"><img class="avatar avatar-md"></div>
            <div class="tile-content">
              <p class="tile-title"><span class="text-bold-600"></span> updated a request</p>
              <p class="tile-subtitle">

              </p>
            </div>
            <div class="tile-action"><span class="label label-warning">Pending</span></div>
          </div>
        </li>
        <li class="">
          <div class="tile tile-centered tile-activity">
            <div class="tile-icon"><img class="avatar avatar-md"></div>
            <div class="tile-content">
              <p class="tile-title"><span class="text-bold-600"></span> updated a request</p>
              <p class="tile-subtitle">

              </p>
            </div>
            <!---->
          </div>
        </li>
      </ul>
    </div>
  </div>
  <modal-stub title="Edit attendance" modal-open="true">
    <div class="form-group"><label class="form-label">Date</label>
      <flatpickr-stub config="[object Object]" disabled="" class="form-input daterange-picker"></flatpickr-stub>
    </div>
    <div class="form-group"><label class="form-label">Attended At</label> <input type="time" disabled="disabled" class="form-input time-picker"></div>
    <div class="form-group"><label class="form-label">Left At</label> <input type="time" disabled="disabled" class="form-input time-picker"></div>
    <div class="form-group"><label class="form-label">Reason</label> <textarea disabled="disabled" class="form-input"></textarea></div>
    <div class="form-group"><label class="form-label">Reject Reason</label> <textarea placeholder="Reject Reason" class="form-input"></textarea></div>
    <div class="modal-footer"><button type="button" class="btn btn-success">
        Approve
      </button> <button type="button" class="btn btn-error">
        Reject
      </button></div>
  </modal-stub>
</div>
`;
